app/src/main/java/org/wikipedia/recommendedcontent/RecommendedContentViewModel.kt

```diff
@@ -0,0 +1,278 @@
+                    RecommendedContentSection.PLACES_NEAR_YOU -> async { loadPlaces() }
+                    RecommendedContentSection.BECAUSE_YOU_READ -> async { loadBecauseYouRead(0) }
+                    RecommendedContentSection.CONTINUE_READING -> async { loadContinueReading() }
+                }
+                recommendedContent.add(content)
+            }
+
+            // merge into one list and shuffle
+            val contentList = recommendedContent.map { it.await() }.flatten().distinct().shuffled()
+
+            _recommendedContentState.value = Resource.Success(contentList)
+        }
+    }
+
+    private suspend fun loadRecentSearches(): List<PageTitle> {
+        return withContext(Dispatchers.IO) {
+            AppDatabase.instance.recentSearchDao().getRecentSearches().map {
+                PageTitle(it.text, wikiSite).apply {
+                    // Put timestamp in description for the delete action.
+                    description = it.timestamp.time.toString()
+                }
+            }.take(RECENT_SEARCHES_ITEMS)
+        }
+    }
+
+    private suspend fun loadFeed(): AggregatedFeedContent {
+        return withContext(Dispatchers.IO) {
+
+            feedContent[wikiSite.languageCode]?.let {
+                return@withContext it
+            }
+
+            val hasParentLanguageCode = !WikipediaApp.instance.languageState.getDefaultLanguageCode(wikiSite.languageCode).isNullOrEmpty()
+            val date = DateUtil.getUtcRequestDateFor(0)
+            var feedContentResponse = ServiceFactory.getRest(wikiSite).getFeedFeatured(date.year, date.month, date.day)
+            if (hasParentLanguageCode) {
+                feedContentResponse.topRead?.let {
+                    val topReadResponse = L10nUtil.getPagesForLanguageVariant(it.articles, wikiSite)
+                    feedContentResponse = AggregatedFeedContent(
+                        tfa = feedContentResponse.tfa,
+                        news = feedContentResponse.news,
+                        topRead = TopRead(it.date, topReadResponse),
+                        potd = feedContentResponse.potd,
+                        onthisday = feedContentResponse.onthisday
+                    )
+                }
+            }
+            // set map to feedContent
+            feedContent[wikiSite.languageCode] = feedContentResponse
+            feedContentResponse
+        }
+    }
+
+    private suspend fun loadTopRead(): List<PageSummary> {
+        return withContext(Dispatchers.IO) {
+            loadFeed().topRead?.articles?.take(5) ?: emptyList()
+        }
+    }
+
+    private suspend fun loadOnThisDay(): List<PageSummary> {
+        return withContext(Dispatchers.IO) {
+            loadFeed().onthisday?.first()?.pages() ?: emptyList()
+        }
+    }
+
+    private suspend fun loadInTheNews(): List<PageSummary> {
+        return withContext(Dispatchers.IO) {
+            loadFeed().news?.first()?.links?.filterNotNull() ?: emptyList()
```

Comment:
This is actually not quite correct:  this gives us all the links from _just the first_ news item. But we actually want to get the _first link from each_ news item.
Something like:
`loadFeed().news?.mapNotNull { it.links.firstOrNull() } ?: emptyList()`

Reply:
Got it. Updated!